### DISCUSSION TITLE
Elkjs Layout does not position nodes correctly in case of multiple source handles in a node

### DISCUSSION THREAD

**Discussion by abdelazkamo** [xyflow/xyflow#4248] [Layouting]
### Describe the Bug

I am using Elkjs example code for my project to layout nodes. but they are not positioned in the correct order in case of multiple handles in a specific node.

### Steps to Reproduce the Bug or Issue

1. create a node with multiple handles
2. connect each handle into a new node
3. apply horizontal alignment
4. see the error

### Expected behavior

The current result is shown in image 1 and the expected result in image 2.

### Screenshots or Videos

![1](https://github.com/xyflow/xyflow/assets/77016461/cc6f8f50-2196-48b1-86d9-3adb7f33d06f)
![2](https://github.com/xyflow/xyflow/assets/77016461/90f95785-2c7e-42de-93c3-b3c1cd9cffff)

### Platform

- OS: [e.g. macOS, Windows, Linux]
- Browser: [e.g. Chrome, Safari, Firefox]
- Version: [e.g. 91.1]

**Comment by moklick** [comment]
This is not an issue with React Flow but a question about how to use Elkjs. Please refer to their documentation https://eclipse.dev/elk/reference.html

You can solve it by adding ports to the elkjs graph nodes and also specify the edges correctly by using the same port ids for `sources` and `targets`. This is how a elkjs graph could look like:

```js 
 const graph = {
    id: 'root',
    layoutOptions,
    children: nodes.map((n) => {
      const sourcePorts = n.data.sources
        ? n.data.sources.map((s, si) => ({
            id: s.sourceHandle,
          }))
        : [{ id: n.id }];

      return {
        id: n.id,
        width: n.width,
        height: n.height,
        ports: [{ id: n.id }, ...sourcePorts],
      };
    }),
    edges: edges.map((e) => ({
      id: e.id,
      sources: [e.sourceHandle || e.source],
      targets: [e.target],
    })),
  };
```

**Comment by abdelazkamo** [comment]
Thank you for your response. I added the ports to the elkjs graph nodes as indicated but that didn't really solve the problem. I may have missed something. I accessed the sources via symbol property because they were not present in the data object. This is what the nodes look like in the graph with the ports added.

![image](https://github.com/xyflow/xyflow/assets/77016461/15f5ba7d-df7c-4a00-9f91-24aa113fbd61)

Please take a look at my code and tell me if I need to adjust anything to make it work.

```
import ELK from 'elkjs/lib/elk.bundled.js';
import React, { useCallback } from 'react';
import { useReactFlow, useStoreApi } from 'reactflow';
import AutoFixHighIcon from '@mui/icons-material/AutoFixHigh';

import 'reactflow/dist/style.css';

const elk = new ELK();

const elkOptions = {
   'elk.algorithm': 'layered',
   'elk.layered.spacing.nodeNodeBetweenLayers': '100',
   'elk.layered.nodePlacement.strategy': 'NETWORK_SIMPLEX',
   'elk.spacing.nodeNode': '80',
   'elk.hierarchyHandling': 'INCLUDE_CHILDREN',
};

const Layout = ({ setNodes, setEdges, edges }) => {
   const { fitView } = useReactFlow();
   const store = useStoreApi();

   const { nodeInternals } = store.getState();
   const nodes = Array.from(nodeInternals).map(([, node]) => node);

   const getLayoutedElements = (nodes, edges, options = {}) => {
      if (nodes.length > 0) {
         const graph = {
            id: 'root',
            layoutOptions: options,
            children: nodes.map((n) => {
               const symbolprop = Object.getOwnPropertySymbols(n);
               const handleBounds = n[symbolprop[0]].handleBounds;
               const sources = handleBounds.source && handleBounds.source;

               const sourcePorts = sources
                  ? sources.map((s, si) => ({
                       id: s.id,
                       order: si,
                    }))
                  : [{ id: n.id, order: 0 }];

               return {
                  ...n,
                  ports: [{ id: n.id }, ...sourcePorts],
               };
            }),

            edges: edges.map((e) => ({
               id: e.id,
               sources: [e.sourceHandle || e.source],
               targets: [e.target],
            })),
         };

         return elk
            .layout(graph)
            .then((layoutedGraph) => {
               return {
                  nodes: layoutedGraph.children.map((node) => ({
                     ...node,
                     position: { x: node.x, y: node.y },
                  })),
                  edges: layoutedGraph.edges,
               };
            })
            .catch(console.error);
      }
   };

   const onLayout = useCallback(
      ({ direction }) => {
         const opts = { 'elk.direction': direction, ...elkOptions };
         const ns = nodes;
         const es = edges;

         getLayoutedElements(ns, es, opts).then(({ nodes: layoutedNodes, edges: layoutedEdges }) => {
            setNodes(layoutedNodes);
            setEdges(layoutedEdges);

            window.requestAnimationFrame(() => fitView());
         });
      },
      [nodes, edges, setNodes, setEdges, fitView],
   );

   return (
      <div>
         <AutoFixHighIcon
            onClick={() => onLayout({ direction: 'RIGHT' })}
            style={{ width: '20px', height: '15px', margin: '2px 1.3px 0 0' }}
         />
      </div>
   );
};

export default Layout;

```

**Comment by moklick** [comment]
We added a new example to show how to configure elkjs to work with multiple handles https://reactflow.dev/examples/layout/elkjs-multiple-handles

**Reply by manansoni77** [reply]
Can this also be achieved in svelte-flow?

**Comment by chetanviamagus** [comment]
@moklick 
https://reactflow.dev/examples/layout/elkjs-multiple-handles
How to change the layout to Vertical. top to bottom instead of left to right.

**Comment by jamalsoueidan** [comment]
I have the same issue, and im not sure if this problem is related to the handle ordering or the height of the node.

**Comment by RomainSF** [comment]
try with BRANDES_KOEPF in BALANCED mode
+ avoidNodes=true and avoidEdges=true

**Reply by jamalsoueidan** [reply]
What is BRANDES_KOEPF ?are you using dagreGraph?

**Comment by nawangwe** [comment]
I have this problem too. I have all the port ids mapped, and included in the edge data, but it still crisscrosses. It only works when the edges are in their correct order in the edge array, which means that the port ordering simply has no effect. I'm using a layered algorithm, with direction of down.

**Comment by albert849** [comment]
> ### Describe the Bug
> 
> I am using Elkjs example code for my project to layout nodes. but they are not positioned in the correct order in case of multiple handles in a specific node.
> 
> ### Your Example Website or App
> 
> _No response_
> 
> ### Steps to Reproduce the Bug or Issue
> 
> 1. create a node with multiple handles
> 
> 2. connect each handle into a new node
> 
> 3. apply horizontal alignment
> 
> 4. see the error
> 
> ### Expected behavior
> 
> The current result is shown in image 1 and the expected result in image 2.
> 
> ### Screenshots or Videos
> 
> ![1](https://github.com/xyflow/xyflow/assets/77016461/cc6f8f50-2196-48b1-86d9-3adb7f33d06f)
> 
> ![2](https://github.com/xyflow/xyflow/assets/77016461/90f95785-2c7e-42de-93c3-b3c1cd9cffff)
> 
> 
> 
> ### Platform
> 
> - OS: [e.g. macOS, Windows, Linux]
> 
> - Browser: [e.g. Chrome, Safari, Firefox]
> 
> - Version: [e.g. 91.1]
> 
> 
> 
> ### Additional context
> 
> _No response_

**Comment by ZahidChandio** [comment]
Anyone looking for the answer, please try below options.
```
const elkOptions = {
  'elk.algorithm': 'layered',
  'elk.direction': 'DOWN',
  'elk.layered.nodePlacement.strategy': 'BRANDES_KOEPF',
  'elk.layered.nodePlacement.bk.fixedAlignment': 'BALANCED',
  'elk.layered.considerModelOrder.portModelOrder': 'true',
  'elk.layered.considerModelOrder.strategy': 'NODES_AND_EDGES',
  'elk.layered.crossingMinimization.forceNodeModelOrder': 'true',
  'elk.layered.compaction.connectedComponents': 'true',
  'elk.layered.compaction.postCompaction.strategy': 'EDGE_LENGTH',
  'elk.layered.cycleBreaking.strategy': 'MODEL_ORDER',
};
```

**Comment by dimesv** [comment]
Hi @moklick,

I am having a few issues with your example: https://reactflow.dev/examples/layout/elkjs-multiple-handles
 
The first issue is that no edges were being displayed. I fixed this by adapting the getLayoutedNodes function and also return the edges. 
Then just underneath setNodes(layoutedNodes); i also set the edges as follows setEdges(layoutedEdges);

This works fine however, the layout does not position nodes correctly in case of multiple target handles in a node. It works well for multiple source handles but not multiple targets. 

Copy the graph code below and confirm that in ELK Editor this works well. https://rtsys.informatik.uni-kiel.de/elklive/json.html?

`{
    "id": "root",
    "layoutOptions": {
        "elk.algorithm": "layered",
        "org.eclipse.elk.direction": "RIGHT",
        "org.eclipse.elk.layered.spacing.edgeNodeBetweenLayers": 40,
        "org.eclipse.elk.spacing.nodeNode": 600,
        "org.eclipse.elk.layered.spacing.nodeNodeBetweenLayers": 1000,
        "org.eclipse.elk.layered.nodePlacement.strategy": "SIMPLE"
    },
    "children": [
        {
            "id": "b",
            "width": 400,
            "height": 200,
            "ports": [
                {
                    "id": "b-s-a",
                    "x": 50,
                    "y": 200,
                    "width": 10,
                    "height": 10,
                    properties: {
                      side: 'EAST',
                    }
                }
            ]
        },
        {
            "id": "d",
            "width": 400,
            "height": 200,
            "ports": [
                {
                    "id": "d-s-a",
                    "x": 50,
                    "y": 200,
                    "width": 10,
                    "height": 10,
                    properties: {
                      side: 'EAST',
                    }
                }
            ]
        },
        {
            "id": "e",
            "width": 400,
            "height": 200,
            "ports": [
                {
                    "id": "e-s-a",
                    "x": 50,
                    "y": -10,
                    "width": 10,
                    "height": 10,
                    properties: {
                      side: 'WEST',
                    }
                },
                {
                    "id": "e-s-b",
                    "x": 100,
                    "y": -10,
                    "width": 10,
                    "height": 10,
                    properties: {
                      side: 'WEST',
                    }
                }
            ]
        }
    ],
    "edges": [
        {
            "layoutOptions": {},
            "id": "edge1",
            "sources": [
                "b-s-a"
            ],
            "targets": [
                "e-s-a"
            ]
        },
        {
            "layoutOptions": {},
            "id": "edge2",
            "sources": [
                "d-s-a"
            ],
            "targets": [
                "e-s-b"
            ]
        },
    ]
}`

<img width="712" height="452" alt="image" src="https://github.com/user-attachments/assets/0fa65c65-0594-4791-ba27-58daef32f8a5" />

This is what I expected. But the settings used above do not seem to work the same way in react flow as you can see from the result below.

<img width="773" height="588" alt="image" src="https://github.com/user-attachments/assets/f113a367-1ab3-4bc2-b68e-4beda9f309b3" />

In react flow, I have only been able to get the same result by applying the port index which is not ideal for me as I would rather have the algorithm plotting the correct position of nodes and edges automatically. I don't think I can control the model to a point where I can set the port index myself so I would rather have this work well just like in the ELK Editor example above.

Thank you in advance,
Jose De Sousa
